crates/client/block_production/src/lib.rs

```diff
@@ -335,20 +287,19 @@ impl<Mempool: MempoolProvider> BlockProductionTask<Mempool> {
             }
         }
 
-        // Add back the unexecuted transactions to the mempool.
-        stats.n_re_added_to_mempool = txs_to_process.len();
-        self.mempool.re_add_txs(txs_to_process);
+        let on_top_of = self.executor.block_state.as_ref().expect(BLOCK_STATE_ACCESS_ERR).state.on_top_of_block_id;
 
-        let on_top_of = self
-            .executor
-            .block_state
-            .as_ref()
-            .expect("Block state can not be None unless we take ownership of it")
-            .state
-            .on_top_of_block_id;
+        // TODO: save visited segments for SNOS.
+        let (state_diff, visited_segments, bouncer_weights) = finalize_execution_state::finalize_execution_state(
+            &executed_txs,
+            &mut self.executor,
+            &self.backend,
+            &on_top_of,
+        )?;
 
-        let (state_diff, _visited_segments, _weights) =
-            finalize_execution_state(&executed_txs, &mut self.executor, &self.backend, &on_top_of)?;
+        // Add back the unexecuted transactions to the mempool.
+        stats.n_re_added_to_mempool = txs_to_process.len();
+        self.mempool.re_add_txs(txs_to_process, executed_txs);
```

Comment:
ouch woah good catch

i think the
```rs
if block_now_full {
  break;
}
```
breaks out the wrong loop! it needs to be a break with label
uuh i'll add a test before fixing that because that's pretty bad that i did not find this earlier

Reply:
no, it's breaking from the correct loop afaik, i misread
this `loop` has two exits: either txs_to_process is empty (mempool has no more transactions for us)
or block_now_full is true, meaning blockifier has returned fewer transactions than what we asked for, meaning bouncer cap has been reached.
this re_add is useful for the later case only, to put the remaining items back into the mempool